src/test/ui/lifetimes/issue-105227.stderr

```diff
@@ -0,0 +1,59 @@
+error[E0700]: hidden type for `impl Iterator<Item = char>` captures lifetime that does not appear in bounds
+  --> $DIR/issue-105227.rs:8:5
+   |
+LL | fn chars0(v :(& str, &str)) -> impl Iterator<Item = char> {
+   |               ----- hidden type `std::iter::Chain<Chars<'_>, Chars<'_>>` captures the anonymous lifetime defined here
+...
+LL |     v.0.chars().chain(v.1.chars())
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+help: to declare that `impl Iterator<Item = char>` captures `'_`, you can consider introducing a named lifetime parameter `'a`
+   |
+LL | fn chars0<'a>(v :(&'a  str, &'a str)) -> impl Iterator<Item = char> + 'a  {
+   |          ++++      ++        ++                                     ++++
+
+error[E0700]: hidden type for `impl Iterator<Item = char>` captures lifetime that does not appear in bounds
+  --> $DIR/issue-105227.rs:8:5
+   |
+LL | fn chars0(v :(& str, &str)) -> impl Iterator<Item = char> {
+   |                      ---- hidden type `std::iter::Chain<Chars<'_>, Chars<'_>>` captures the anonymous lifetime defined here
+...
+LL |     v.0.chars().chain(v.1.chars())
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+help: to declare that `impl Iterator<Item = char>` captures `'_`, you can consider introducing a named lifetime parameter `'a`
+   |
+LL | fn chars0<'a>(v :(&'a  str, &'a str)) -> impl Iterator<Item = char> + 'a  {
+   |          ++++      ++        ++                                     ++++
```

Comment:
It'd be amazing if we could deduplicate these E0700s based on the primary span, precisely so that we can turn this test into a `// run-rustfix` test (and reduce verbosity to boot. We might even go further and not just hide the later one but rather incorporate *all* the spans pointing at the params capturing the hidden lifetimes.

This should be follow up work.

Reply:
We already have a buffering mechanism, but it occurs at the `Diagnostic` level. I wonder if we should move these to operate on a higher level enum or struct so that we have more info available to merge multiple errors into one.

https://github.com/rust-lang/rust/blob/a5cd3bde95da07589b7a3a0cee35ec9b390c72aa/compiler/rustc_borrowck/src/diagnostics/region_errors.rs#L215-L226

Seeing this I wonder if the deduplication logic is not triggering because of the secondary spans only.

Reply:
The deduplication logic in `fn emit_diagnostic(&mut self, diagnostic: &mut Diagnostic)` will not be triggered because the two spans in `span_lables` field of two `diagnostic` records are different:
```
Diagnostic {.., span: MultiSpan { primary_spans: [issue-105227.rs:8:5: 8:35 (#0)], span_labels: [(issue-105227.rs:5:15: 5:20 (#0), Str("hidden type `std::iter::Chain<Chars<'_>, Chars<'_>>` captures the anonymous lifetime defined here"))] }..}
```
vs
```
Diagnostic {.., span: MultiSpan { primary_spans: [issue-105227.rs:8:5: 8:35 (#0)], span_labels: [(issue-105227.rs:5:22: 5:26 (#0), Str("hidden type `std::iter::Chain<Chars<'_>, Chars<'_>>` captures the anonymous lifetime defined here"))] }..}
```
Although the `primary_spans` and `label` field of `span_labels` are the same. 
We may take out the span in `span_labels` when doing hash for `diagnostic` records to mark them duplicates for the case,  but need to take care of any side effects.

Reply:
I've tried the following code change to deduplicate these E0700s based on the (primary_span, named_ty and named_key) tuple, seems working as expected without side effects:
```rust
    pub(crate) fn report_region_errors(&mut self, nll_errors: RegionErrors<'tcx>) {
        // Iterate through all the errors, producing a diagnostic for each one. The diagnostics are
        // buffered in the `MirBorrowckCtxt`.

        let mut outlives_suggestion = OutlivesSuggestionBuilder::default();
+       let mut last_unexpected_hidden_region: Option<(Span, Ty<'_>, ty::OpaqueTypeKey<'tcx>)> = None;
        for nll_error in nll_errors.into_iter() {
            match nll_error {
... ...
                RegionErrorKind::UnexpectedHiddenRegion { span, hidden_ty, key, member_region } => { 
                    let named_ty = self.regioncx.name_regions(self.infcx.tcx, hidden_ty); 
                    let named_key = self.regioncx.name_regions(self.infcx.tcx, key); 
                    let named_region = self.regioncx.name_regions(self.infcx.tcx, member_region); 
+                   if last_unexpected_hidden_region != Some((span, named_ty, named_key)) {
                        self.buffer_error(unexpected_hidden_region_diagnostic(
                            self.infcx.tcx,
                            span,
                            named_ty,
                            named_region,
                            named_key,
                        ));
+                       last_unexpected_hidden_region = Some((span, named_ty, named_key));
+                   }
```
All tests passed and including this test with `run-rustfix` enabled. Is this a reasonable solution for fixing the issue?

Reply:
The only thing I'd change is in the `else` branch to still *construct* the error but call `.delay_as_bug()` on it. That way, if there's some fluke caused by a future refactor where no errors are emitted we still won't emit an invalid binary.

Reply:
Thanks for the advice. Added `.delay_as_bug()` in the `else` branch.